File: python/memory_pool/agent.py
import os
import json
from pathlib import Path
from openai import AsyncOpenAI
from typing import AsyncIterator, Callable
from .pool import MemoryPool
from .types import MemoryEntry, Turn
from .worker import MemoryWorker, extract_keywords
from .llm import stream_chat_complete, chat_complete

BASE_URL = "https://generativelanguage.googleapis.com/v1beta/openai"
DEFAULT_MAIN_MODEL = "gemma-4-31b-it"
DEFAULT_MEMORY_MODEL = "gemma-4-26b-a4b-it"


class MemoryAgent:
    """High-level agent combining LLM chat with async memory compression."""

    def __init__(
        self,
        api_key: str | None = None,
        main_model: str = DEFAULT_MAIN_MODEL,
        memory_model: str = DEFAULT_MEMORY_MODEL,
        base_url: str = BASE_URL,
        max_raw_turns: int = 4,
        min_turns_to_compress: int = 2,
        max_pool_entries: int = 1000,
        context_window: int | None = None,
        pool_ratio: float = 0.15,
        inject_budget: int = 800,
        pool_file: str | None = None,
        on_event: Callable[[str, dict], None] | None = None,
    ):
        self.main_model = main_model
        self.max_raw_turns = max_raw_turns
        self.inject_budget = inject_budget
        self.pool_file = Path(pool_file) if pool_file else None
        self.on_event = on_event or (lambda kind, data: None)

        self.client = AsyncOpenAI(
            api_key=api_key or os.environ.get("GEMINI_API_KEY", ""),
            base_url=base_url,
        )
        self.pool = MemoryPool(
            max_entries=max_pool_entries,
            context_window=context_window,
            pool_ratio=pool_ratio,
        )
        self.worker = MemoryWorker(
            pool=self.pool,
            client=self.client,
            memory_model=memory_model,
            min_turns_to_compress=min_turns_to_compress,
            on_event=self._handle_event,
        )

        self._turns: list[Turn] = []
        self._turn_index = 0
        self._tasks: set = set()   # strong refs to prevent GC cancellation

    def _spawn(self, coro) -> None:
        import asyncio
        task = asyncio.create_task(coro)
        self._tasks.add(task)
        task.add_done_callback(self._tasks.discard)

    def _handle_event(self, kind: str, data: dict) -> None:
        self.on_event(kind, data)
        if kind == "compress_done" and self.pool_file:
            self._spawn(self._save_pool())

    async def load_pool(self) -> None:
        if not self.pool_file or not self.pool_file.exists():
            return
        try:
            entries = json.loads(self.pool_file.read_text())
            self.pool.load([MemoryEntry.from_dict(e) for e in entries])
            self.on_event("pool_state", {"restored": self.pool.size, "source": str(self.pool_file)})
        except Exception as err:
            self.on_event("compress_error", {"error": f"pool load failed: {err}"})

    async def _save_pool(self) -> None:
        if not self.pool_file:
            return
        try:
            self.pool_file.write_text(
                json.dumps([e.to_dict() for e in self.pool.get_all()], indent=2)
            )
        except Exception:
            pass

    def _build_messages(self, user_message: str, memory_text: str) -> list[dict]:
        recent = self._turns[-self.max_raw_turns:]
        messages = [{"role": "system", "content": "You are a helpful assistant."}]
        for t in recent[:-1]:
            messages.append({"role": t.role, "content": t.content})
        content = f"## Conversation Memory\n{memory_text}\n\n---\n\n{user_message}" if memory_text else user_message
        messages.append({"role": "user", "content": content})
        return messages

    async def chat_stream(self, user_message: str) -> AsyncIterator[str]:
        user_turn = Turn(role="user", content=user_message, index=self._turn_index)
        self._turn_index += 1
        self._turns.append(user_turn)
        self.on_event("turn_user", {"index": user_turn.index, "chars": len(user_message)})

        mem_entries = await self.worker.get_memory(user_message, self.inject_budget)
        memory_text = "\n".join(f"[{e.keyword}] {e.content}" for e in mem_entries)

        messages = self._build_messages(user_message, memory_text)
        self.on_event("context_built", {
            "rawTurns": min(len(self._turns), self.max_raw_turns),
            "memoryEntries": len(mem_entries),
            "totalMessages": len(messages),
        })

        full_text = ""
        async for chunk in stream_chat_complete(self.client, self.main_model, messages):
            full_text += chunk
            yield chunk

        asst_turn = Turn(role="assistant", content=full_text, index=self._turn_index)
        self._turn_index += 1
        self._turns.append(asst_turn)
        self.on_event("turn_assistant", {"index": asst_turn.index, "chars": len(full_text)})

        self.worker.add_turn(user_turn)
        self.worker.add_turn(asst_turn)

    async def chat(self, user_message: str) -> str:
        result = ""
        async for chunk in self.chat_stream(user_message):
            result += chunk
        return result

File: bun-npm/src/agent.ts
import OpenAI from 'openai'
import { MemoryPool } from './pool'
import { MemoryWorker, extractKeywords } from './worker'
import { streamChatComplete, chatComplete } from './llm'
import type { Turn, MemoryEntry, AppEvent, MemoryAgentOptions } from './types'

const DEFAULT_BASE_URL = 'https://generativelanguage.googleapis.com/v1beta/openai'
const DEFAULT_MAIN_MODEL = 'gemma-4-31b-it'
const DEFAULT_MEMORY_MODEL = 'gemma-4-26b-a4b-it'

export class MemoryAgent {
  private client: OpenAI
  private pool: MemoryPool
  private worker: MemoryWorker
  private turns: Turn[] = []
  private turnIndex = 0
  private poolFile?: string
  private emit: (event: AppEvent) => void

  readonly maxRawTurns: number
  readonly injectBudget: number
  readonly mainModel: string

  constructor(options: MemoryAgentOptions = {}) {
    this.mainModel = options.mainModel ?? DEFAULT_MAIN_MODEL
    this.maxRawTurns = options.maxRawTurns ?? 4
    this.injectBudget = options.injectBudget ?? 800
    this.poolFile = options.poolFile
    this.emit = options.onEvent ?? (() => {})

    this.client = new OpenAI({
      apiKey: options.apiKey ?? process.env.GEMINI_API_KEY ?? '',
      baseURL: options.baseURL ?? DEFAULT_BASE_URL,
    })

    this.pool = new MemoryPool({
      maxEntries: options.maxPoolEntries ?? 1000,
      contextWindow: options.contextWindow,
      poolRatio: options.poolRatio,
    })
    this.worker = new MemoryWorker(
      this.pool,
      this.client,
      options.memoryModel ?? DEFAULT_MEMORY_MODEL,
      options.minTurnsToCompress ?? 2,
      (event) => {
        this.emit(event)
        if (event.kind === 'compress_done' && this.poolFile) {
          this.savePool()
        }
      },
    )
  }

  async loadPool(): Promise<void> {
    if (!this.poolFile) return
    try {
      const file = Bun.file(this.poolFile)
      if (await file.exists()) {
        const entries = await file.json() as MemoryEntry[]
        this.pool.load(entries)
        this.emit({ kind: 'pool_state', ts: Date.now(), data: { restored: entries.length, source: this.poolFile } })
      }
    } catch (err) {
      this.emit({ kind: 'compress_error', ts: Date.now(), data: { error: `pool load: ${err}` } })
    }
  }

  private async savePool(): Promise<void> {
    if (!this.poolFile) return
    try { await Bun.write(this.poolFile, JSON.stringify(this.pool.getAll(), null, 2)) } catch { /* non-critical */ }
  }

  private buildMessages(userMessage: string, memoryText: string): OpenAI.Chat.ChatCompletionMessageParam[] {
    const recent = this.turns.slice(-this.maxRawTurns)
    const messages: OpenAI.Chat.ChatCompletionMessageParam[] = [
      { role: 'system', content: 'You are a helpful assistant.' },
    ]
    for (const t of recent.slice(0, -1)) messages.push({ role: t.role, content: t.content })
    const content = memoryText ? `## Conversation Memory\n${memoryText}\n\n---\n\n${userMessage}` : userMessage
    messages.push({ role: 'user', content })
    return messages
  }

  async *chatStream(userMessage: string): AsyncGenerator<string> {
    const userTurn: Turn = { role: 'user', content: userMessage, index: this.turnIndex++ }
    this.turns.push(userTurn)
    this.emit({ kind: 'turn_user', ts: Date.now(), data: { index: userTurn.index, chars: userMessage.length } })

    const memEntries = this.worker.getMemory(userMessage, this.injectBudget)
    const memoryText = memEntries.map(e => `[${e.keyword}] ${e.content}`).join('\n')

    const messages = this.buildMessages(userMessage, memoryText)
    this.emit({ kind: 'context_built', ts: Date.now(), data: {
      rawTurns: Math.min(this.turns.length, this.maxRawTurns),
      memoryEntries: memEntries.length,
      totalMessages: messages.length,
    }})

    let fullText = ''
    for await (const chunk of streamChatComplete(this.client, this.mainModel, messages)) {
      fullText += chunk
      yield chunk
    }

    const asstTurn: Turn = { role: 'assistant', content: fullText, index: this.turnIndex++ }
    this.turns.push(asstTurn)
    this.emit({ kind: 'turn_assistant', ts: Date.now(), data: { index: asstTurn.index, chars: fullText.length } })

    this.worker.addTurn(userTurn)
    this.worker.addTurn(asstTurn)
  }

  async chat(userMessage: string): Promise<string> {
    let result = ''
    for await (const chunk of this.chatStream(userMessage)) result += chunk
    return result
  }

  getPoolEntries(): MemoryEntry[] { return this.pool.getAll() }
  getPoolSize(): number { return this.pool.size }
}
